dali/pipeline/operators/reader/loader/lmdb.h

```diff
@@ -107,64 +189,50 @@ class LMDBLoader : public Loader<CPUBackend, Tensor<CPUBackend>> {
     }
 
     tensor.SetMeta(meta);
-    tensor.Resize({static_cast<Index>(value_.mv_size)});
+    tensor.Resize({static_cast<Index>(value.mv_size)});
     std::memcpy(tensor.raw_mutable_data(),
-                reinterpret_cast<uint8_t*>(value_.mv_data),
-                value_.mv_size*sizeof(uint8_t));
+                reinterpret_cast<uint8_t*>(value.mv_data),
+                value.mv_size * sizeof(uint8_t));
   }
 
  protected:
-  Index SizeImpl() override {
-    return lmdb_size_;
-  }
+  Index SizeImpl() override { return size_array_[db_path_.size()]; }
 
   void PrepareMetadataImpl() override {
-    // Create the db environment, open the passed DB
-    CHECK_LMDB(mdb_env_create(&mdb_env_), db_path_);
-    auto mdb_flags = MDB_RDONLY | MDB_NOTLS | MDB_NOLOCK;
-    CHECK_LMDB(mdb_env_open(mdb_env_, db_path_.c_str(), mdb_flags, 0664), db_path_);
-
-    // Create transaction and cursor
-    CHECK_LMDB(mdb_txn_begin(mdb_env_, NULL, MDB_RDONLY, &mdb_transaction_), db_path_);
-    CHECK_LMDB(mdb_dbi_open(mdb_transaction_, NULL, 0, &mdb_dbi_), db_path_);
-    CHECK_LMDB(mdb_cursor_open(mdb_transaction_, mdb_dbi_, &mdb_cursor_), db_path_);
-    lmdb_size_ = lmdb::LMDB_size(mdb_transaction_, mdb_dbi_, db_path_);
-
-    // Optional: debug printing
-    lmdb::PrintLMDBStats(mdb_transaction_, mdb_dbi_, db_path_);
-
+    size_array_.resize(db_path_.size() + 1);
+    size_array_[0] = 0;
+    mdb_.resize(db_path_.size());
+    for (size_t i = 0; i < db_path_.size(); i++) {
+      mdb_[i].Open(db_path_[i], i);
+      size_array_[i + 1] = size_array_[i] + mdb_[i].GetSize();
+    }
     Reset(true);
   }
 
  private:
   void Reset(bool wrap_to_shard) override {
     // work out how many entries to move forward to handle sharding
-    current_index_ = start_index(shard_id_, num_shards_, Size());
-    bool ok = lmdb::SeekLMDB(mdb_cursor_, MDB_FIRST, &key_, &value_, db_path_);
-    DALI_ENFORCE(ok, "lmdb::SeekLMDB to the beginning failed");
-
     if (wrap_to_shard) {
-      for (size_t i = 0; i < current_index_; ++i) {
-        bool ok = lmdb::SeekLMDB(mdb_cursor_, MDB_NEXT, &key_, &value_, db_path_);
-        DALI_ENFORCE(ok, "lmdb::SeekLMDB to position " + to_string(current_index_) + " failed");
-      }
+      current_index_ = start_index(shard_id_, num_shards_, Size());
+    } else {
+      current_index_ = 0;
     }
+    Index file_index, local_index;
+    MapIndexToFile(current_index_, &file_index, &local_index);
+
+    mdb_[file_index].SeekByIndex(local_index);
   }
   using Loader<CPUBackend, Tensor<CPUBackend>>::shard_id_;
   using Loader<CPUBackend, Tensor<CPUBackend>>::num_shards_;
 
-  MDB_env* mdb_env_ = nullptr;
-  MDB_cursor* mdb_cursor_ = nullptr;
-  MDB_dbi mdb_dbi_;
-  MDB_txn* mdb_transaction_ = nullptr;
-  size_t current_index_;
-  Index lmdb_size_;
+  std::vector<IndexedLMDB> mdb_;
+
+  Index current_index_;
 
-  // values
-  MDB_val key_, value_;
+  std::vector<Index> size_array_;
```

Comment:
maybe something like
```suggestion
  std::vector<Index> offsets_;
```